File: Value.cpp

#include "Value.hpp"

#include "State.hpp"

void Value0::go() {
  state->action = state->action0;
  state->go_1();
}

void Value0::print(std::ostream& os) const {
  os << "ZERO";
}

void Value1::go() {
  state->action = state->action1;
  state->go_1();
}

void Value1::print(std::ostream& os) const {
  os << "ONE";
}

std::ostream& operator<<(std::ostream& os, const Value* value) {
  value->print(os);
  return os;
}

File: js/untyped/Value.js

function value0Action(call = true) {
  if (call) this.state.action0.doCell();
  else return [this.state.action0, 'doCell'];
}

function value0String() {
  return "ZERO";
}

function Value0() {
  if (new.target) throw new TypeError();
  const block = {
    toString: value0String,
    state: null,
    doAction: value0Action,
  };
  return Object.seal(block);
}

function value1Action(call = true) {
  if (call) this.state.action1.doCell();
  else return [this.state.action1, 'doCell'];
}

function value1String() {
  return "ONE";
}

function Value1() {
  if (new.target) throw new TypeError();
  const block = {
    toString: value1String,
    state: null,
    doAction: value1Action,
  };
  return Object.seal(block);
}

module.exports = {Value0, Value1};
